### ISSUE TITLE
More training data is needed

### ISSUE THREAD

**Issue by alizat** [Dainanahan/IDG-DREAM-Drug-Kinase-Binding-Prediction-Challenge#12]
The first dataset that has been collected now lies in the folder:
**data/dataset1**

The first dataset was collected using the code that is in:
**dataset1.R**

The first dataset collected was fine. When used to train our prediction models, the best performing model was used to obtain predictions for the test set. The predictions were then submitted, and our standing in the leaderboards was reasonable for a 1st submission. However, we want to do better now, and the most obvious way to do that is by training our prediction models using more training data. 

The target variable is supposed to be in the form of pKd values. While records with such values are available in the data downloaded from [Drug Target Commons](https://drugtargetcommons.fimm.fi/), they are very few in number (a few hundred out of millions of records).

Earlier, it was observed that there are records containing binding affinity info in the form of Kd values in units of nanomolars (or NM for short). The formula for converting the Kd NM values into pKd values is:

***pKd = -log10(Kd * 10^-9)***

These records (i.e. those with toupper(standard_type)=='KD' and standard_units=='NM') are to have their values (from the 'standard_value' column) converted to pKd values using the above formula. After that, these records will be added to those of the first dataset to form the second dataset (which is the objective).

Here is what needs to be done:

- Make a folder structure similar to that in data/dataset1. In other words, make sure these folders exist before writing any code:
    - data/dataset2/
    - data/dataset2/train/
    - data/dataset2/train/MOL/
    - data/dataset2/train/SMILES/
    - data/dataset2/test/
- Make a copy of **dataset1.R** and call it  -->  **dataset2.R**
- Modify the first line of **dataset2.R** as:
    - path <- 'data/dataset2/'
- We will create a dataset that will contain the same exact data as the first one (generated by **dataset1.R**), but we'll add to the new dataset the records whose Kd values will be converted to pKd values
    - After obtaining the records containing the pKd values (already existing code), get the records with Kd NM values
    - Filter the full dtc_data using the following criteria:
        - standard_type must be 'KD' (convert to uppercase before matching with 'KD')
        - standard_unit must be 'NM'
        - standard_relation must be '='
        - !is.na(compound_id) & !is.na(target_id) & !is.na(standard_value)
    - After all the above filtering keep only the columns: compound_id, target_id, standard_value
    - Rename standard_value to pkd
    - Before transforming Kd values to pKd ones, keep only records with Kd values in the range [0.01, 100000]  (inclusive)
    - Transform values using the formula mentioned earlier above. 
        - That is assuming Kd=x, then pKd=-log10(10^-9 * kds)
    - If there are multiple records containing the same compound-target pair but with different pKd values, get the mean --> the code for doing this is already in the code of **dataset2.R**
    - With the new set of records obtained, what remains is to obtain the features for these records. The code for obtaining the features should work without any problems whatsoever. Do note that there a couple of manual steps involved in obtaining the targets' features (getting the protein sequences from UniProt & getting the features from PROFEAT). Instructions for manually doing these steps are explained in detail in the code already. While these steps are also to be done for the test set, note that test set is still the same as when the first dataset was generated and the features for the test set have already been previously computed.

**Comment by alizat** [comment]
Problems encountered/Lessons Learned:
- ALWAYS ALWAYS ALWAYS do the following checks regularly
    - check for NAs
    - check for empty strings (do a trim() first then check)
    - check for duplicates when there shouldn't be any (e.g. every distinct compound has to have its own unique SMILES/InchiKey)
- Writing to many files (i.e. more than a few hundred) in a for loop is almost always a bad idea and should be avoided whenever possible or else R gives unexpected behavior. If it is inevitable, proceed with extreme caution and keep watching progress like a hawk!
    - If possible, write all info to a single file
    - It is better to use functions that handle the file creation/opening/closing automatically such as readr::write_lines() and cat()
- The Rcpi::extractDrugAIO() ***sometimes*** produces the error below. The probability of this appearing increases when the number of SMILES strings being processed is very large (i.e. thousands or more). The solution is to repeat executing the function until it stops producing the error message below:
    - `Error in eval.desc(molecules, descNames, verbose = !silent) : molecule must be an IAtomContainer`

**Comment by alizat** [comment]
What's been done:
- Target features were obtained using the Rcpi package (Previously, they were obtained using the PROFEAT server). The reason Rcpi was used is that the PROFEAT was out of order at the time. The set of descriptors obtained using Rcpi is very similar to that obtained by PROFEAT for the previous dataset.
- Compound features were obtained using the Mordred (Python) package as it is more convenient and works as intended. Before obtaining features, the compounds with overly long SMILES strings (>125) were removed because no SMILES strings in the test set are that long anyway. Number of features here (1426) is bigger than that of the features obtained by Rcpi for the previous dataset (165).
- Size of the dataset obtained (27256 observations) is much larger than that of the previous dataset (500+ observations).

What was found:
- Rcpi is unreliable when it comes to generating features for many SMILES strings that are stored in the same file. Specifically, when there are 10 or more SMILES in the file and Rcpi is asked to generate features for them, expect trouble! 
- While the test set was generated and uploaded successfully to this GitHub repo, the training set was too big to upload. They may be generated using lines 742-762 in **dataset2.R**.
